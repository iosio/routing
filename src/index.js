import {isArray, isObj} from "@iosio/util";
import {obi} from "@iosio/obi";

export const createRouting = () => {
    let w = window,
        toValue = (mix) => {
            if (!mix) return '';
            var str = decodeURIComponent(mix);
            if (str === 'false') return false;
            if (str === 'true') return true;
            return (+str * 0 === 0) ? (+str) : str;
        },
        getParams = (str) => {
            let tmp, k, out = {}, indi;
            str = str || w.location.search;
            indi = str.indexOf("?");
            if (indi < 0) return;
            str = str.substr(indi + 1);
            let arr = str.split('&');
            while (tmp = arr.shift()) {
                tmp = tmp.split('=');
                k = tmp.shift();
                if (out[k] !== void 0) out[k] = [].concat(out[k], toValue(tmp.shift()));
                else out[k] = toValue(tmp.shift());
            }
            return out;
        },
        stringifyParams = (obj) => {
            var enc = encodeURIComponent, k, i, tmp, str = '';
            for (k in obj) {
                if ((tmp = obj[k]) !== void 0) {
                    if (isArray(tmp)) {
                        for (i = 0; i < tmp.length; i++) {
                            str && (str += '&');
                            str += enc(k) + '=' + enc(tmp[i]);
                        }
                    } else {
                        str && (str += '&');
                        str += enc(k) + '=' + enc(tmp);
                    }
                }
            }
            return '?' + str;
        },
        getLocation = () => {
            let {pathname, search} = w.location;
            return {url: pathname + search, pathname, search, params: getParams()}
        },
        setUrl = (path, search, type) => {
            search = isObj(search) ? stringifyParams(search) : search;
            w.history[type + 'State'](null, null, path + search)
        },

        initLoc = getLocation(),

        {url: _lastUrl, pathname: _lastPathname} = initLoc;

    let routing = obi({
        // pre-pending with $ ignores updates when setting them
        $lastUrl: _lastUrl,
        $lastPathname: _lastPathname,
        $lastType: 'initial',
        ...initLoc,
        getParams,
        stringifyParams,
        getLocation,
        route(path, search, type) {
            type = type || 'push';
            path = path || location.pathname;
            search = search || '';
            const {pathname, url} = getLocation();
            if (type !== 'replace') (_lastUrl = url, _lastPathname = pathname);
            setUrl(path, search, type);
            type === 'replace'
                ? setTimeout(() => updateLocation({type}))
                : updateLocation({type});
        }
    });

    routing.routerSwitch = ({root, pathMap, noMatch}) => {
        let next = null, toLast = false,
            {pathname, $lastPathname, $lastUrl, $lastType, url} = routing,
            noChange = url === $lastUrl;
        noMatch = noMatch || '/';
        if (root) next = pathMap['/' + pathname.split('/')[1]] || pathMap[noMatch];
        else if (pathMap[pathname]) next = pathMap[pathname];
        else if ($lastPathname !== pathname && pathMap[$lastPathname]) {
            routing.route($lastUrl, location.search, 'replace');
            next = pathMap[$lastPathname];
            toLast = true;
        } else if (noMatch && pathMap[noMatch]) {
            routing.route(noMatch, location.search, 'replace');
            next = pathMap[noMatch];
        }
        return {next, toLast, noChange, replacedLast: $lastType === 'replace'};
    };

    var updateLocation = (e) => {
        routing.$merge({
            ...getLocation(),
            $lastUrl: e.type === 'popstate' ? routing.url : _lastUrl,
            $lastPathname: e.type === 'popstate' ? routing.pathname : _lastPathname,
            $lastType: e.type
        })
    };

    w.addEventListener("popstate", updateLocation);

    return routing;
};

export const routing = createRouting();

