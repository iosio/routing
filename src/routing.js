import React from 'react';
import {Subscription} from "@iosio/utils/lib/subscription";
import {typeOf} from '@iosio/utils/lib/typeOf';
import {getParams, stringifyParams, flattenChildren, preparePath} from "./utils";
// npm publish --access public
const ROUTERS = [];
let h = window.history;
const sub = Subscription();
const notify = () => sub.notify(getLoc());
export const subscribe = (fn) => sub.subscribe(fn);
export const unsubscribe = (fn) => sub.unsubscribe(fn);
export const goBack = () => h.go(-1);
export const goForward = () => h.go(1);

export const getLoc = () => ({
    pathname: window.location.pathname,
    search: window.location.search,
    params: getParams()
});

export const makeRoutable = (_path, _search) => {
    let path = _path || getLoc().pathname;
    let search = _search ? (typeOf(_search) === 'object' ? stringifyParams(_search) : _search) : '';
    const hasSlashAlready = path.length && (path.charAt(path.length - 1) === "/");
    const slash = hasSlashAlready ? "" : (search ? "/" : "");
    return {path, search, url: path + slash + search}
};

export const getUrl = () => makeRoutable(getLoc().pathname, getLoc().search).url;

let lastUrl = getUrl();
export const setUrl = (url, type = "push") => {
    h[type + 'State'](null, null, url || '');
    lastUrl = url;
};

const canRoute = (path) => {
    for (let i = ROUTERS.length; i--;) if (ROUTERS[i].canRoute(path)) return true;
    return false;
};

/** Tell all router instances to handle the given URL.  */
function routeTo(path) {
    for (let i = 0; i < ROUTERS.length; i++) {
        ROUTERS[i].routeTo(path)
    }
    notify();
}

const tryRoute = (_path, _params, type = 'push') => {
    const {path, url} = makeRoutable(_path, _params);
    if (url === lastUrl) return;
    if (canRoute(preparePath(path))) {
        setUrl(url);
        routeTo(path);
    }
};

/**
 * handles familiar arguments of type {pathname:'/', search: '?id=3'}
 * or pass the path and search in separate arguments ('/', {id: 3})
 * @returns {undefined} - returns nothing
 */
export const goTo = function () {
    const arg1 = arguments.length <= 0 ? undefined : arguments[0];
    typeOf(arg1) === 'object'
        ? tryRoute(arg1.pathname, arg1.search)
        : tryRoute.apply(undefined, arguments)
};

let inited = false;
const initEventListeners = () => {
    if (inited) return;
    inited = true;
    window.addEventListener("popstate", () => {
        routeTo(getLoc().pathname);
    });
};

export class Router extends React.Component {
    constructor(props) {
        super(props);
        initEventListeners();
        this.state = {CurrentRoute: null};
        this.accessible = [];
        this.excluded = [];
        this.pathMap = {};

        this.lastPath = preparePath(true);

        const hasChildren = this.getChildren().length;
        if (hasChildren > 0) {
            this.register();
            if (!this.canRoute()) {
                let p = this.getRedirectPath();
                setUrl(p, 'replace');
                this.lastPath = p;
            }
            this.state = {
                CurrentRoute: this.pathMap[this.lastPath] || null
            };

        }

        ROUTERS.push(this);
    }

    routeTo = (path) => {
        const pn = preparePath(path);
        this.setState({
            CurrentRoute: this.pathMap[pn] || null
        });
        this.lastPath = pn;
    };

    getRedirectPath = () => {
        const {noMatchRedirectTo, root} = this.props;
        //assuming we cant route to the existing base path

        let redirectTo = '/';
        //is there is a default path provided? and can we go there?
        if (noMatchRedirectTo && this.canRoute(noMatchRedirectTo)) {
            redirectTo = noMatchRedirectTo;
            // console.log('going to "noMatchRedirectTo"');
        }
        //if not see if we can go back to the last path
        //only if it doesnt match the current - to prevent infinite loops
        else if ((preparePath(true) !== this.lastPath) && this.canRoute(this.lastPath)) {
            redirectTo = this.lastPath;
            // console.log("no 'noMatchRedirectTo' prop exists. will go to last Path");
        }
        //if we cant go back to the last path
        //see if we can go back to the root if its provided
        else if (root && this.canRoute(root)) {
            // console.log("going to root");
            redirectTo = root;
        } else {
            // console.log('hoping for a "/" path ');
        }
        //if neither one of those work, lets hope they have a root "/" path set up
        //otherwise this.getMatchingComponent will return null;
        return redirectTo;
    };

    canRoute = (path) => this.accessible.includes(path || preparePath(true));

    getChildren = (children) => flattenChildren(children || this.props.children);

    register = () => {
        const {isAuthenticated} = this.props;
        this.accessible = [];
        this.excluded = [];
        this.pathMap = {};

        this.getChildren().forEach((child) => {
            const props = child ? (child.length ? child[0].props : child.props) : {};
            const {authOnly, notAuth, restricted, path} = props;
            const pathName = preparePath(path);
            let excluded = false;
            if (pathName) {
                if ((!isAuthenticated && authOnly) || (isAuthenticated && notAuth)) {
                    excluded = true;
                }
                if (restricted && !restricted.canView) {
                    excluded = true;
                }
                if (!excluded) {
                    this.accessible.push(pathName);
                    this.pathMap[pathName] = child;
                } else {
                    this.excluded.push(pathName);
                }
            }
        });
    };

    componentDidUpdate(prevProps) {
        const {isAuthenticated, onLogoutPath, onLoginPath, userChange, onUserChangePath} = this.props;

        this.register();

        if (prevProps.userChange !== userChange) {

            if (onUserChangePath) {
                setUrl(onUserChangePath, 'replace');
                this.routeTo(onUserChangePath);
                notify();
            }
        }
        if (prevProps.isAuthenticated !== isAuthenticated) {
            if (!isAuthenticated) {
                if (onLogoutPath && this.excluded.includes(preparePath(true))) {
                    setUrl(onLogoutPath, 'replace');
                    this.routeTo(onLogoutPath);
                    notify();
                }
            } else if (onLoginPath) {
                setUrl(onLoginPath);
                this.routeTo(onLoginPath);
                notify();
            }
        }
    }

    componentWillUnmount() {
        ROUTERS.splice(ROUTERS.indexOf(this), 1);
    }

    render() {
        return this.state.CurrentRoute;
    }
}

export class Link extends React.Component {
    constructor(props) {
        super(props);
        this.state = getLoc();
        this.unsubscribe = sub.subscribe((location) => {
            this.setState({...location});
        })
    }

    componentWillUnmount() {
        this.unsubscribe && this.unsubscribe();
    }

    render() {
        const {render, children, to, params, href, ...rest} = this.props;
        const place = to || href;
        return (
            <a href={place}
               onClick={(e) => {
                   if (to) {
                       if (e.stopImmediatePropagation) e.stopImmediatePropagation();
                       if (e.stopPropagation) e.stopPropagation();
                       goTo(to, params);
                       e.preventDefault();
                   }
               }} {...rest}>
                {render ? render({...this.state}) : children}
            </a>
        )
    }
}



