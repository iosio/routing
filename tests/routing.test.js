import React, {useState} from 'react'
import renderer from 'react-test-renderer';
import {Router, Link, goTo, getLoc, subscribe} from "../src";
import {makeRoutable, getUrl, setUrl} from "../src/routing";
import {render, fireEvent, cleanup, queryByAttribute} from 'react-testing-library';

import ReactDOM from 'react-dom';
import 'jest-dom/extend-expect'

const getById = queryByAttribute.bind(null, 'id');


describe('Routing', () => {

    it('getLoc should return the correct object', () => {

        expect(getLoc()).toMatchObject({
            pathname: '/',
            search: '',
            params: false
        })
    });

    it('makeRoutable should construct an object containing a path, stringified search as well as a combined pathname and search as url', () => {

        const expectedSearch = '?id=3&user=Joe%20Dirt';

        expect(makeRoutable('/', expectedSearch)).toMatchObject({
            path: '/',
            search: expectedSearch,
            url: '/' + expectedSearch
        });

        expect(makeRoutable('/', {id: 3, user: 'Joe Dirt'})).toMatchObject({
            path: '/',
            search: expectedSearch,
            url: '/' + expectedSearch
        });

        expect(makeRoutable('/hello', {id: 3, user: 'Joe Dirt'})).toMatchObject({
            path: '/hello',
            search: expectedSearch,
            url: '/hello/' + expectedSearch
        });

        expect(makeRoutable('/hello/', {id: 3, user: 'Joe Dirt'})).toMatchObject({
            path: '/hello/',
            search: expectedSearch,
            url: '/hello/' + expectedSearch
        });

    });

    it('setUrl should set the correct url, getUrl should return it, subscribe should pic up the changes', () => {
        setUrl('/hello');
        expect(getUrl()).toBe("/hello");

        setUrl('/hello/?id=3');
        expect(getUrl()).toBe("/hello/?id=3");
    });


});

/*

        const cb = jest.fn();

        let locObj = {};

        const unsubscribe = subscribe((location)=>{
            locObj = location;
            cb(location)
        });
        expect(cb).toHaveBeenCalledWith(locObj)
 */


const Page = (props) => {
    return (
        <div className={'page'}>
            <label>{props.name}</label>

            <br/>
            <br/>
            props: {JSON.stringify(props, null, 4)}

        </div>
    )
};

let canView = false;
const routes = [
    {
        path: '/',
        Component: Page,
        name: 'Home Page'
    },
    {
        path: '/detail',
        Component: Page,
        name: 'Detail'
    },
    {
        path: '/authOnly',
        Component: Page,
        authOnly: true,
        name: 'Authenticated only!'
    },
    {
        path: '/restricted',
        Component: Page,
        authOnly: true,
        restricted: {canView},
        name: 'Admin only!'
    },
    {
        path: '/login',
        Component: Page,
        notAuth: true,
        name: 'Login Page',
    }
];

const renderRoutes = () => routes.map(({Component, ...rest}, i) => (
    <Component key={i} {...rest}/>
));

afterEach(() => {
    cleanup();
});

describe('Router functionality', () => {


    it('renders without crashing', () => {
        setUrl('/');

        const tree = renderer.create(
            <Router/>
        );


        expect(tree).toMatchSnapshot();


        tree.unmount();

    });


    it('renders the correct routes and replaces the history if no path exists before mounting', () => {
        setUrl('/hello');

        const tree = renderer.create(
            <Router>
                {renderRoutes()}
            </Router>
        );


        expect(tree).toMatchSnapshot();
        expect(getUrl()).toBe('/')

        tree.unmount();
        setUrl('/');
    });


    it('renders the login page when navigated to and when notAuth is present on the page props', () => {
        expect(getUrl()).toBe('/')


        const tree = renderer.create(
            <Router>
                {renderRoutes()}
            </Router>
        );

        goTo('/login')

        expect(tree).toMatchSnapshot();


        tree.unmount();
        setUrl('/');

    });


    // beforeEach(() => {
    //     container = document.createElement('div');
    //     document.body.appendChild(container);
    // });
    //
    // afterEach(() => {
    //     document.body.removeChild(container);
    //     container = null;
    // });


    it('Goes to the appropriate route when authState Changes', () => {

        setUrl('/');
        expect(getUrl()).toBe('/')

        const onLoginPath = '/authOnly';

        const rendered = jest.fn();

        const Page = (props) => {
            rendered(props.name);
            return (
                <div>
                    {props.aux && props.aux()}
                    <label>{props.name}</label>
                </div>
            )
        };

        const App = () => {
            const [hasAuth, setAuth] = useState(false);

            return (
                <React.Fragment>
                    <button id={'toggleLogin'} onClick={() => setAuth(!hasAuth)}>loginBtn</button>

                    <Router isAuthenticated={hasAuth}
                            onLogoutPath={'/login'}
                            onLoginPath={onLoginPath}>

                        <Page path={'/'} name={'Home'} notAuth
                              aux={() => <button id={'goTo'} onClick={() => goTo('/login')}>go to login page</button>}
                        />

                        <Page path={'/login'} name={'loginPage'} notAuth/>

                        <Page path={onLoginPath} authOnly name={'authOnly'}/>

                    </Router>
                </React.Fragment>
            )
        };

        const {container} = render(<App/>);

        fireEvent.click(getById(container, "goTo"));

        expect(rendered).toHaveBeenCalledWith('loginPage');

        expect(getUrl()).toBe('/login');

        fireEvent.click(getById(container, "toggleLogin"));

        expect(rendered).toHaveBeenCalledWith('authOnly');

        expect(getUrl()).toBe('/authOnly')

        fireEvent.click(getById(container, "toggleLogin"));

        expect(getUrl()).toBe('/login')

    })

});


