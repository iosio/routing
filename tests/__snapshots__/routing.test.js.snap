// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router functionality renders the correct routes and replaces the history if no path exists before mounting 1`] = `
<div
  className="page"
>
  <label>
    Home Page
  </label>
  <br />
  <br />
  props: 
  {
    "path": "/",
    "name": "Home Page"
}
</div>
`;

exports[`Router functionality renders the login page when navigated to and when notAuth is present on the page props 1`] = `
<div
  className="page"
>
  <label>
    Login Page
  </label>
  <br />
  <br />
  props: 
  {
    "path": "/login",
    "notAuth": true,
    "name": "Login Page"
}
</div>
`;

exports[`Router functionality renders without crashing 1`] = `null`;
